### PR TITLE
[SPARK-34164][SQL] Improve write side varchar check to visit only last few tailing spaces

### PR DESCRIPTION
**Pull request by yaooqinn**
### What changes were proposed in this pull request?

For varchar(N), we currently trim all spaces first to check whether the remained length exceeds, it not necessary to visit them all but at most to those after N.

### Why are the changes needed?

improve varchar performance for write side
### Does this PR introduce _any_ user-facing change?

no
### How was this patch tested?

benchmark and existing ut

### CODE REVIEW

**Comment by yaooqinn** [comment]
cc @cloud-fan @maropu @HyukjinKwon thanks

File: common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 1]
```diff
@@ -683,6 +683,16 @@ public UTF8String trimRight() {
     return copyUTF8String(0, e);
   }
 
+  /**
+   * Return the length of string stripped the tailing space after `limit`.
+   */
+  public int canTrimRightTo(int numOfSpaces) {
```
Comment (cloud-fan) [inline common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java:689]: it's a bit weird that a method with the name `canTrimRightTo` returns int not boolean.

Comment (yaooqinn) [reply common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java]: renamed to `effectiveLength`m WDYT?

File: common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 2]
```diff
@@ -683,6 +683,16 @@ public UTF8String trimRight() {
     return copyUTF8String(0, e);
   }
 
+  /**
+   * Return the length of string stripped at most `numOfSpaces` tailing spaces
+   */
+  public int effectiveLength(int numOfSpaces) {
```
Comment (cloud-fan) [inline common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java:689]: how about `public UTF8String trimTrailingSpaces(int numSpaces)`? I don't think we need to save the string creating for the error case, as perf doesn't matter much for that case.

Comment (yaooqinn) [reply common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java:689]: Considering that we need to calculate the new string's numChars again for error raising.

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/util/CharVarcharCodegenUtils.java [hunk 3]
```diff
@@ -47,19 +47,23 @@ public static UTF8String charTypeReadSideCheck(UTF8String inputStr, int limit) {
...
+    } else {
+      int numChars = inputStr.numChars();
+      if (numChars <= limit) {
+        return inputStr;
+      } else {
+        // Trailing spaces do not count in the length check. We need to retain the trailing spaces
+        // (truncate to length N), as there is no read-time padding for varchar type.
+        int tailSpacesMayNeedTrim = numChars - limit;
+        int lengthAfterTrim = inputStr.effectiveLength(tailSpacesMayNeedTrim);
+        if (lengthAfterTrim == limit) {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/util/CharVarcharCodegenUtils.java:61]: This isn't right. `lengthAfterTrim` is the length of bytes, while `limit` is the length of chars.

Comment (yaooqinn) [reply sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/util/CharVarcharCodegenUtils.java:61]: oh...

File: common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java (hunk 2 above)
Comment (yaooqinn) [reply common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java:689]: ah good point!

File: common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java [hunk 4]
```diff
@@ -683,6 +683,16 @@ public UTF8String trimRight() {
     return copyUTF8String(0, e);
   }
 
+  /**
+   * Return the string stripped at most `numOfSpaces` tailing spaces
```
Comment (cloud-fan) [inline common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java:687]: `Returns the string that is trimmed at most ``numSpaces`` tailing spaces`

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/util/CharVarcharCodegenUtils.java [hunk 5]
```diff
@@ -47,19 +47,23 @@ public static UTF8String charTypeReadSideCheck(UTF8String inputStr, int limit) {
...
+    if (inputStr == null) {
       return null;
+    } else {
+      int numChars = inputStr.numChars();
+      if (numChars <= limit) {
+        return inputStr;
+      } else {
+        // Trailing spaces do not count in the length check. We need to retain the trailing spaces
+        // (truncate to length N), as there is no read-time padding for varchar type.
+        int tailSpacesMayNeedTrim = numChars - limit;
```
Comment (cloud-fan) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/util/CharVarcharCodegenUtils.java:59]: `maxAllowedNumTailSpaces`

File: sql/core/benchmarks/CharVarcharBenchmark-results.txt [hunk 6]
```diff
@@ -6,85 +6,130 @@ Java HotSpot(TM) 64-Bit Server VM 1.8.0_251-b08 on Mac OS X 10.16
...
+write varchar with length 100                      2466           2486          23          4.1         246.6       0.9X
+
+
+================================================================================================
+Char Varchar Write Side Perf w Tailing Spaces
+================================================================================================
+
+Java HotSpot(TM) 64-Bit Server VM 1.8.0_251-b08 on Mac OS X 10.16
+Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
+Write with length 20, hasSpaces: true:    Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
```
Comment (cloud-fan) [inline sql/core/benchmarks/CharVarcharBenchmark-results.txt:97]: I think we don't need to print `hasSpaces: true` now.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/CharVarcharBenchmark.scala [hunk 7]
```diff
@@ -112,9 +112,16 @@ object CharVarcharBenchmark extends SqlBasedBenchmark {
...
 
-    runBenchmark("Char Varchar Write Side Perf") {
+    runBenchmark("Char Varchar Write Side Perf w/o Tailing Spaces") {
       for (len <- range) {
-        writeBenchmark(N / 10, len, len >= 80)
+        writeBenchmark(N / 10, len, hasSpaces = false)
+      }
+    }
+
+    runBenchmark("Char Varchar Write Side Perf w Tailing Spaces") {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/CharVarcharBenchmark.scala:121]: can we do the same for the read benchmark?

**Comment by yaooqinn** [comment]
@cloud-fan I’ve addressed the comments，can you check again

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**